src/kbd-slide.c

```diff
@@ -0,0 +1,217 @@
+    }
+
+    libevdev_free(dev);
+    close(fd);
+    return FALSE;
+}
+
+/**
+ * Find and set proper input device for slide keypad monitoring
+ * @return  TRUE if found, FALSE otherwise
+ */
+static gboolean
+kbd_slide_find_evdev()
+{
+    DIR *dir = NULL;
+    struct dirent *direntry = NULL;
+    gboolean result = FALSE;
+
+    if ((dir = opendir(DEV_INPUT_PATH)) == NULL) {
+        perror(__func__);
+        return FALSE;
+    }
+
+    for (direntry = readdir(dir);
+         (direntry != NULL && telldir(dir));
+         direntry = readdir(dir))
+    {
+        gchar *dev_path = NULL;
+        if (strncmp(direntry->d_name, EVENT_FILE_PREFIX,
+                    strlen(EVENT_FILE_PREFIX)) != 0)
+        {
+            continue;
+        }
+
+        dev_path = g_strconcat(DEV_INPUT_PATH, "/", direntry->d_name, NULL);
+
+        if (kbd_slide_set_input_dev(dev_path)) {
+            result = kbd_slide_add_handler();
+            g_free(dev_path);
+            break;
+        }
+
+        g_free(dev_path);
+    }
+
+    /* Report, but ignore, errors when closing directory */
+    if (closedir(dir) == -1)
+        perror(__func__);
+
+    return result;
+}
+
+/**
+ * Keypad slide state changes handler
+ */
+static gboolean
+kbd_slide_handler(GIOChannel *src, GIOCondition cond, gpointer data)
+{
+    struct input_event ev;
+    int rc;
+
+    for (;;) {
+        rc = libevdev_next_event(kbd_slide_dev, LIBEVDEV_READ_FLAG_NORMAL, &ev);
+
+        if (rc == LIBEVDEV_READ_STATUS_SYNC) {
+            /* We are out of sync, fix it */
+            do {
+                rc = libevdev_next_event(kbd_slide_dev,
+                                         LIBEVDEV_READ_FLAG_SYNC, &ev);
+            } while (rc == LIBEVDEV_READ_STATUS_SYNC);
+
+            /* Re-read current keyboard slide state */
+            gboolean state = libevdev_get_event_value(kbd_slide_dev,
+                                                      EV_SW, SW_KEYPAD_SLIDE);
+            inform_slide_keyboard(state);
+            continue;
+        }
+
+        if (rc == LIBEVDEV_READ_STATUS_SUCCESS) {
+            if (libevdev_event_is_code(&ev, EV_SW, SW_KEYPAD_SLIDE)) {
+                inform_slide_keyboard(ev.value);
+            }
+            continue;
+        }
+
+        if (rc == -EAGAIN) {
+            /* There's no more input data */
+            return TRUE;
+        }
+
+        /* An error occured or we are out of sync, stop handling */
+        return FALSE;
+    }
+}
+
+static gboolean
+kbd_slide_add_handler()
+{
+    kbd_slide_iochan = g_io_channel_unix_new(libevdev_get_fd(kbd_slide_dev));
+    if (kbd_slide_iochan == NULL)
+        return FALSE;
+
+    kbd_slide_src_id = g_io_add_watch(kbd_slide_iochan, G_IO_IN,
+                                      kbd_slide_handler, NULL);
+    return TRUE;
+}
+
+void
+kbd_slide_monitor_start()
+{
+    if (kbd_slide_dev != NULL) {
+        /* Already started */
+        return;
+    }
+
+    if (!kbd_slide_find_evdev()) {
```

Comment:
What would be the initial state reported? Is libevdev going to report anything before there is a state change?

Reply:
The initial state of slide is set here: https://github.com/maemo-leste/ke-recv/blob/38ff8e22d5a28e42d733930e03a2c11f7c21f5d7/src/kbd-slide.c#L68

Reply:
ok, fair enough